pkg/codegen/hcl2/rewrite_convert.go

```diff
@@ -90,6 +90,9 @@ func rewriteConversions(x model.Expression, to model.Type) (model.Expression, bo
 	case *model.IndexExpression:
 		x.Key, _ = rewriteConversions(x.Key, x.KeyType())
 	case *model.ObjectConsExpression:
+		if v := resolveDiscriminatedUnions(to, x); v != nil {
```

Comment:
Hmm... I think this is the wrong spot to do that. E.g., I see that I miss conversions for tuples containing objects. I'll try to think of a better one - let me know if you know the answer!